Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,383 @@
+        // but the body itself is throwing
+        let result = await withTaskGroup(of: ChildTaskResult.self, returning: Result<Void, Error>.self) { group in
+            // First we have to register our signals.
+            let unixSignals = await UnixSignalsSequence(trapping: self.configuration.gracefulShutdownSignals)
+
+            // This is the task that listens to signals
+            group.addTask {
+                for await signal in unixSignals {
+                    return .signalCaught(signal)
+                }
+
+                return .signalSequenceFinished
+            }
+
+            // This is the task that listens to manual graceful shutdown
+            group.addTask {
+                for await _ in gracefulShutdownStream {
+                    return .gracefulShutdownCaught
+                }
+
+                return .gracefulShutdownFinished
+            }
+
+            // This is an optional task that listens to graceful shutdowns from the parent task
+            if let _ = TaskLocals.gracefulShutdownManager {
+                group.addTask {
+                    for await _ in AsyncGracefulShutdownSequence() {
+                        return .gracefulShutdownCaught
+                    }
+
+                    return .gracefulShutdownFinished
+                }
+            }
+
+            // We have to create a graceful shutdown manager per service
+            // since we want to signal them individually and wait for a single service
+            // to finish before moving to the next one
+            var gracefulShutdownManagers = [GracefulShutdownManager]()
+            gracefulShutdownManagers.reserveCapacity(self.services.count)
+
+            for (index, service) in self.services.enumerated() {
+                self.logger.debug(
+                    "Starting service",
+                    metadata: [
+                        self.configuration.logging.serviceKey: "\(service)",
+                    ]
+                )
+
+                let gracefulShutdownManager = GracefulShutdownManager()
+                gracefulShutdownManagers.append(gracefulShutdownManager)
+
+                // This must be addTask and not addTaskUnlessCancelled
+                // because we must run all the services for the below logic to work.
+                group.addTask {
+                    return await TaskLocals.$gracefulShutdownManager.withValue(gracefulShutdownManager) {
+                        do {
+                            try await service.run()
+                            return .serviceFinished(service: service, index: index)
+                        } catch {
+                            return .serviceThrew(service: service, index: index, error: error)
+                        }
+                    }
+                }
+            }
+
+            precondition(gracefulShutdownManagers.count == self.services.count, "We did not create a graceful shutdown manager per service")
```

Comment:
assert better choice?

Reply:
A precondition here is better IMO. If we this condition isn't `true` we made a mistake in our implementation. We could also remove this precondition but it just to make sure everything is setup correctly before we proceed.